Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
-        let guessDigits = Int(Double(Words.Element.bitWidth) * log10(2))
-        let guessMagnitude = pow(10, Double(guessDigits))
-
-        if let magnitudeAsSelf = Self(exactly: guessMagnitude) {
-            return (guessDigits, magnitudeAsSelf)
-        }
-
-        // Alas `Self` is smaller than UInt, so fall back to a truly generic - but slower - algorithm to find the results.  This is because BinaryIntegers - unlike e.g. FixedWidthIntegers - don't provide APIs for questions like "what is the maximum bit width of Self?".
-
-        var count = 1
-        var value: Words.Element = 1
-
-        while true {
-            var (nextValue, overflowed) = value.multipliedReportingOverflow(by: 10)
-
-            // Words.Element might be wider than the actual type, e.g. if Self is UInt8 (and UInt is not).  The magnitude is limited by the smallest of the two.
-            if !overflowed && nil == Self(exactly: nextValue) {
-                overflowed = true
-            }
-
-            if !overflowed || .zero == nextValue {
-                count += 1
-            }
-
-            if overflowed {
-                return (count - 1, Self(value))
-            }
+/// Writes `word` in "Numeric string" format (https://speleotrove.com/decimal/daconvs.html)  to
+/// the trailing edge of `buffer`, then returns the least significant `buffer` index written to.
+///
+/// - Parameters:
+///   - word: An unsigned binary integer's only word.
+///   - buffer: A buffer with enough memory to accommodate the format.
+///
+private func formTrailingNumericStringRepresentationForUnsignedInteger(
+word: UInt, into buffer: Slice<UnsafeMutableBufferPointer<UInt8>>) -> Int {
+    var magnitude = word
+    var index = buffer.endIndex
+    
+    repeat {
+        precondition(index > buffer.startIndex)
+        buffer.base.formIndex(before: &index)
+        
+        let remainder: UInt
+        (magnitude, remainder) = magnitude.quotientAndRemainder(dividingBy: 10)
+        buffer.base[index] = UInt8(ascii: "0") &+ UInt8(truncatingIfNeeded: remainder)
+    }   while magnitude != 0 as UInt
+    
+    return index as Int
+}
 
-            value = nextValue
-        }
+/// Forms the `quotient` of dividing the `dividend` by the `divisor`, then returns the `remainder`.
+///
+/// - Parameters:
+///   - dividend: An unsigned binary integer's words.
+///   - divisor:  An unsigned binary integer's only word.
+///
+/// - Returns: The `remainder`, which is a value in the range of `0 ..< divisor`.
+///
+private func formQuotientWithRemainderForUnsignedInteger(
+words dividend: Slice<UnsafeMutableBufferPointer<UInt>>, dividingBy divisor: UInt) -> UInt {
+    var remainder = UInt.zero
+    
+    for index in (dividend.startIndex ..< dividend.endIndex).reversed() {
+        (dividend.base[index], remainder) = divisor.dividingFullWidth((high: remainder, low: dividend.base[index]))
```

Comment:
It’s a shame that this results in the use of `udivti3` from compiler-rt instead of a simple `divq` instruction (on x86-64; presumably there’s an equivalent instruction on AArch64).  Evidently the compiler doesn’t believe that the result will never overflow (`udivti3` does the extra work of checking for and trapping on overflow, which is a waste here).

Perhaps there's some way to convince the compiler that the checks are unnecessary?  Overflow should be impossible in practice, by the inherent constraints of its use here.

Reply:
That's an astute observation! I'd love to know if there's a better way to do it. I have an open issue in Numberick ([#59](https://github.com/oscbyspro/Numberick/issues/59)) about division being slower than expected in some cases. I wrote a note about udivmodti4 taking up most of the time in Instruments, but I did not know how to proceed beyond that. It's a bit funky that it performs oveflow checking when it truncates the result anyways:

```swift
print(UInt(123).dividingFullWidth((high: 122, low: UInt.max))) // (quotient: UInt.max, remainder: 122)
print(UInt(123).dividingFullWidth((high: 123, low: 00000000))) // (quotient: 00000000, remainder: 000)
```

Reply:
I was mistaken about `__udivti3` re. the overflow case.  It's doesn't actually trap, it just truncates, as you observed.  That's what one gets for trusting StackOverflow. 😝

For completeness… it's essentially using [`__udivmodti4`](https://github.com/apple/llvm-project/blob/76a94e518e019fa3766b8a95d61f46cc474695d1/compiler-rt/lib/builtins/udivmodti4.c#L100C24-L100C36) because `__udivti3` is basically an alias; it's two instructions: one to zero out r8 to effect a `NULL` for the out parameter that reports the remainder, and then a jump to `__udivmodti4` (which incidentally is a waste of space since `__udivmodti4` is the _very_ next procedure, so it'd flow into it automatically without the `jmp`).

`__udivmodti4` is a little complicated, both at the C level and in the resulting assembly (67 instructions, on x86-64, and branchy).  It does use `divq`, ultimately (it's actually hard-coded to on x86-64, via inline assembly).  But you still have the overhead in the function call and unnecessary checks to use other paths - it really would benefit from being inlined.  I don't know how to make the compiler inline it, though.

Which raises the question:  why on earth isn't it just using `__udivmodti4` directly, since we _do_ care about the remainder here.  It appears that it's using the result of the division to calculate the remainder manually:

```assembly
cmp  r12, rdi ; Check that the index is > 0.  On first entry to this loop it's endIndex.
jle  loc_19eb ; Basically, crash if index is <= 0.
lea  rbx, qword [r12-1] ; Decrement index of current word in `words`.
mov  r13, qword [r15+r12*8-8] ; Load next word from `words` argument.
mov  rdi, r13 ; low half of 128-bit argument "a" (dividend) for __udivti3.  rsi is high half, and is initially set to zero but is the remainder in subsequent iterations.
mov  rdx, r14 ; low half of 128-bit argument "b" (divisor) for __udivti3.
xor  ecx, ecx ; high half of argument "b".  Top 32 bits are known zero from earlier in the function.
call __udivti3 ; rax = a / b
mov  rdi, qword [rbp+var_38] ; words.count.
mov  rcx, rax
mul  r14 ; Quotient * divisor (the function argument, not the argument to __udivti3.
mov  rsi, r13 ; The aforementioned next word from `words`.
sub  rsi, rax ; Remainder = current word - (Quotient * divisor)
mov  qword [r15+r12*8-8], rcx ; Save the quotient over the current word in `words`.
mov  r12, rbx ; Just shifting the current index around registers.
cmp  rdi, rbx ; Check that index is != words.count.
jne  loc_19a0 ; Repeat this whole block while it isn't.
```

I don't see how that could be faster than using `__udivmodti4`, since the latter calculates the remainder no matter what - it merely drops it silently if its 'rem' out argument is `NULL`, as it is in the use via `__udivti3`.  Weird.  Seems like an optimiser bug.